lib/iris/experimental/ugrid/utils.py

```diff
@@ -0,0 +1,322 @@
+            if i_dim == mesh_dim:
+                # mesh dim : look for index coords (by name).
+                fulldim = full_mesh_cube.coords(
+                    name_or_coord=index_coord_name, dimensions=(i_dim,)
+                )
+                regdim = region_cube.coords(
+                    name_or_coord=index_coord_name, dimensions=(i_dim,)
+                )
+            else:
+                # non-mesh dims : look for dim-coords (only)
+                fulldim = full_mesh_cube.coords(
+                    dim_coords=True, dimensions=(i_dim,)
+                )
+                regdim = region_cube.coords(
+                    dim_coords=True, dimensions=(i_dim,)
+                )
+
+            if fulldim:
+                (fulldim,) = fulldim
+                full_dimname = fulldim.name()
+                fulldim_metadata = metadata_no_varname(fulldim)
+            if regdim:
+                (regdim,) = regdim
+                reg_dimname = regdim.name()
+                regdim_metadata = metadata_no_varname(regdim)
+
+            err = None
+            # N.B. checks for mesh- and non-mesh-dims are different.
+            if i_dim != mesh_dim:
+                # i_dim == mesh_dim :  checks for non-mesh dims.
+                if fulldim and not regdim:
+                    err = (
+                        f"{reg_cube_str} has no dim-coord for dimension "
+                        "{i_dim}, to match the 'full_mesh_cube' dimension "
+                        f'"{full_dimname}".'
+                    )
+                elif regdim and not fulldim:
+                    err = (
+                        f'{reg_cube_str} has a dim-coord "{reg_dimname}" for '
+                        f"dimension {i_dim}, but 'full_mesh_cube' has none."
+                    )
+                elif regdim != fulldim:
+                    err = (
+                        f'{reg_cube_str} has a dim-coord "{reg_dimname}" for '
+                        f"dimension {i_dim}, which does not match that "
+                        f"of 'full_mesh_cube', \"{full_dimname}\"."
+                    )
+            else:
+                # i_dim == mesh_dim :  different rules for this one
+                if not regdim:
+                    # Must have an index coord on the mesh dimension
+                    err = (
+                        f'{reg_cube_str} has no "{index_coord_name}" coord on '
+                        f"the mesh dimension (dimension {mesh_dim})."
+                    )
+                elif fulldim and regdim_metadata != fulldim_metadata:
+                    # May *not* have full-cube index, but if so it must match
+                    err = (
+                        f"{reg_cube_str} has an index coord "
+                        f'"{index_coord_name}" whose ".metadata" does not '
+                        "match that on 'full_mesh_cube' :  "
+                        f"{regdim_metadata} != {fulldim_metadata}."
+                    )
```

Comment:
No I don't think there's anything to check in that case.
We allow the `mesh_cube` to not have an index coord (since it would have trival values).
But in that case, we can _only_ check this coord it against the other index-coords -- which is what happens next.